File: examples/java/HelloPlugin/src/main/java/falcon/examples/hello/commands/GiveDiamondCommand.java
package falcon.examples.hello.commands;

import falcon.api.Entity;
import falcon.api.Item;
import falcon.api.PermissionDefault;
import falcon.api.Player;
import falcon.api.Plugin;
import falcon.api.command.CommandContext;
import falcon.api.command.CommandPermission;

import java.util.List;
import java.util.Optional;

public final class GiveDiamondCommand {
    public static final String DIAMOND = "minecraft:diamond";
    public static final String PERMISSION = "hello.command.give-diamond";

    private GiveDiamondCommand() {
    }

    public static void registerTo(Plugin plugin) {
        plugin.permissions().add(PERMISSION, PermissionDefault.OPERATOR);
        plugin.commands().add("give-diamond", "Gives you a shiny diamond", "/give-diamond [count]",
                GiveDiamondCommand::execute, CommandPermission.ANY);
    }

    private static boolean execute(CommandContext context) {
        if (!context.hasPermission(PERMISSION)) {
            context.reply("You are not allowed to use this command");
            return true;
        }

        Optional<Player> player = context.player();
        if (player.isEmpty()) {
            context.reply("Only players can receive diamonds");
            return true;
        }

        int count = 1;
        if (!context.arguments().isEmpty()) {
            try {
                count = Integer.parseUnsignedInt(context.arguments().get(0));
            } catch (NumberFormatException exception) {
                return false;
            }
        }

        Optional<Item> created = Item.create(DIAMOND, count);
        if (created.isEmpty()) {
            context.reply("Could not create the diamond");
            return true;
        }

        try (Item diamond = created.get()) {
            diamond.setCustomName("Shiny Diamond");
            diamond.setLore(List.of("A gift from HelloPlugin"));
            if (!player.get().inventory().give(diamond)) {
                Entity entity = player.get().entity();
                entity.level().dropItem(entity.position(), diamond);
            }
        }

        context.reply("Here you go!");
        return true;
    }
}

File: examples/native/HelloPlugin/src/Commands/GiveDiamondCommand.cpp
#include "HelloPlugin/Commands/GiveDiamondCommand.h"

#include <cstdint>
#include <exception>
#include <optional>
#include <string>

void GiveDiamondCommand::registerTo(const falcon::Plugin &plugin) {
    plugin.permissions().add(PERMISSION, falcon::PermissionDefault::Operator);

    plugin.commands().add("give-diamond", "Gives you a shiny diamond", "/give-diamond [count]",
                          [](falcon::CommandContext &context) {
                              if (!context.hasPermission(PERMISSION)) {
                                  context.reply("You are not allowed to use this command");
                                  return true;
                              }

                              const std::optional<falcon::Player> player = context.player();
                              if (!player) {
                                  context.reply("Only players can receive diamonds");
                                  return true;
                              }

                              uint32_t count = 1;
                              if (!context.arguments().empty()) {
                                  try {
                                      count = static_cast<uint32_t>(std::stoul(context.arguments()[0]));
                                  } catch (const std::exception &) {
                                      return false;
                                  }
                              }

                              std::optional<falcon::Item> diamond = falcon::Item::create(DIAMOND, count);
                              if (!diamond) {
                                  context.reply("Could not create the diamond");
                                  return true;
                              }

                              diamond->setCustomName("Shiny Diamond");
                              diamond->setLore({"A gift from HelloPlugin"});
                              if (!player->inventory().give(*diamond))
                                  player->entity().level().dropItem(player->entity().position(), *diamond);

                              context.reply("Here you go!");
                              return true;
                          }, falcon::CommandPermission::Any);
}
